From ad403bf9b1c1932b54adeee43ec93a2d64b68871 Mon Sep 17 00:00:00 2001
From: Leonid Yegoshin <Leonid.Yegoshin@imgtec.com>
Date: Thu, 12 Sep 2013 13:02:36 -0500
Subject: [PATCH 071/105] MIPS: removal of X bit in page tables for HEAP/BSS.

Patch removes eXecute bit in page tables for HEAP/BSS.
It boosts performance because page marked X is _flushed_
each time after COW/swap from cache even for cache coherent systems
in Harvard architecture (!cpu_has_ic_fills_f_dc).
But even Java JIT uses now a special memory region and sets
execute protection for code explicitly.

Patch also sets eXecute Inhibit protection of HEAP/BSS
on CPUs which supports RIXI like proAptiv.

Signed-off-by: Leonid Yegoshin <Leonid.Yegoshin@imgtec.com>
Signed-off-by: Steven J. Hill <Steven.Hill@imgtec.com>
---
 arch/mips/include/asm/page.h |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/arch/mips/include/asm/page.h b/arch/mips/include/asm/page.h
index f6be474..7db0a8a 100644
--- a/arch/mips/include/asm/page.h
+++ b/arch/mips/include/asm/page.h
@@ -202,7 +202,9 @@ extern int __virt_addr_valid(const volatile void *kaddr);
 #define virt_addr_valid(kaddr)						\
 	__virt_addr_valid((const volatile void *) (kaddr))
 
-#define VM_DATA_DEFAULT_FLAGS	(VM_READ | VM_WRITE | VM_EXEC | \
+#define VM_DATA_DEFAULT_FLAGS   (VM_READ | VM_WRITE | \
+				 VM_MAYREAD | VM_MAYWRITE | VM_MAYEXEC)
+#define VM_STACK_DEFAULT_FLAGS   (VM_READ | VM_WRITE | VM_EXEC | \
 				 VM_MAYREAD | VM_MAYWRITE | VM_MAYEXEC)
 
 #define UNCAC_ADDR(addr)	((addr) - PAGE_OFFSET + UNCAC_BASE)
-- 
1.7.1

